🐛 Fix panic when OpenStackCluster.Status.Network is nil in HCP scenarios

bnallapeta opened the pull request:
**What this PR does / why we need it**:
Fixes a nil pointer panic in OpenStackMachineReconciler when OpenStackCluster.Status.Network is nil, which occurs in Hosted Control Plane scenarios. The controller now gracefully handles missing cluster network by checking for nil before access and returning a terminal error instead of panicking. Also adds comprehensive HCP E2E test suite.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2380 

**TODOs**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR. -->

- [ ] squashed commits
- if necessary:
  - [ ] includes documentation
  - [ ] adds unit tests

/hold

bnallapeta commented:
@EmilienM @mdbooth @lentzi90 

Marking it as Draft as we are working on the e2e tests. This is the approach being taken for e2e. Let me know your thoughts.

1. Spawn up a kind cluster -> turn it into a management cluster by installing CAPI/CAPO on it
2. Deploy a k8s cluster on OpenStack using this kind cluster. Let's call this Cluster A
3. Turn Cluster A into management cluster by installing CAPI/CAPO on it
4. Deploy a k8s cluster in the hosted control plane method using kamaji provider using cluster A. Let's call this Cluster B
5. Run the actual e2e tests to test out the panic case on Cluster B


Right now, we are somewhere in the 3rd/4th step and are facing the below challenges:

First, when we specify our Kamaji as a control plane provider in e2e_conf.yaml and name it as kamaji this error pops up:

```
[FAILED] The e2e test config file is not valid
Expected success, but got an error:
    <*errors.fundamental | 0xc000c00228>:
    invalid argument: invalid config: control-plane-provider should be named kubeadm
```

Second, if we try to bypass and name this kamaji as kubeadm, kamaji doesn't get installed as such and later on, we face issues with finding Kamaji CRDs

```
[FAILED] in [It] - /root/cluster-api-provider-openstack/test/e2e/suites/hcp/hcp_helpers.go:148 @ 07/31/25 11:01:25.32 < Exit [It] should create and manage HCP-capable cluster @ 07/31/25 11:01:25.32 (7m12.465s) << Timeline [FAILED] Timed out after 180.000s. Expected success, but got an error: <*apiutil.ErrResourceDiscoveryFailed | 0xc000123428>: unable to retrieve the complete list of server APIs: [kamaji.clastix.io/v1alpha1](http://kamaji.clastix.io/v1alpha1): no matches for [kamaji.clastix.io/v1alpha1](http://kamaji.clastix.io/v1alpha1), Resource= { { Group: "[kamaji.clastix.io](http://kamaji.clastix.io/)", Version: "v1alpha1", }: <*meta.NoResourceMatchError | 0xc0012143c0>{ PartialResource: { Group: "[kamaji.clastix.io](http://kamaji.clastix.io/)", Version: "v1alpha1", Resource: "", }, }, } In [It] at: /root/cluster-api-provider-openstack/test/e2e/suites/hcp/hcp_helpers.go:148 @ 07/31/25 11:01:25.32
```


Would really appreciate your help on this to move forward.

cc @orkhan-os

mdbooth reviewed (commented):
Appreciate this is just a draft, but I had a quick look over it anyway.

bnallapeta commented on controllers/openstackmachine_controller.go:
```diff
@@ -588,7 +584,18 @@ func (r *OpenStackMachineReconciler) getOrCreateMachineServer(ctx context.Contex
+		// If no cluster network is available AND the machine spec did not define any ports with a network, we cannot choose a network.
+		if defaultNetworkID == "" && len(openStackMachine.Spec.Ports) == 0 {
+			return nil, capoerrors.Terminal(infrav1.InvalidMachineSpecReason, "no network configured: cluster network is missing and machine spec does not define ports with a network")
+		}
```
Done. Code is updated.

bnallapeta replied:
Done.

bnallapeta commented on controllers/openstackmachine_controller_test.go:
```diff
@@ -158,27 +155,67 @@ func TestOpenStackMachineSpecToOpenStackServerSpec(t *testing.T) {
 		tt := tests[i]
 		t.Run(tt.name, func(t *testing.T) {
-			spec := openStackMachineSpecToOpenStackServerSpec(tt.spec, identityRef, tags, "", userData, &openStackCluster.Status.WorkerSecurityGroup.ID, openStackCluster.Status.Network.ID)
+			defaultNetID := ""
+			if openStackCluster.Status.Network != nil {
+				defaultNetID = openStackCluster.Status.Network.ID
+			}
```
Done.

bnallapeta replied:
Done.

bnallapeta commented on controllers/openstackmachine_controller_test.go:
```diff
@@ -174,12 +172,99 @@ func TestOpenStackMachineSpecToOpenStackServerSpec(t *testing.T) {
 			},
 		},
+		{
+			name: "Cluster network nil, machine defines port network and overrides SG",
```
Addressed.

lentzi90 reviewed (commented):
Am I understanding correctly that this changes the behavior for security groups on OpenStackMachines? We cannot sneak in that change as a bug fix (🐛 ). If we want to change it (and I am not sure we do), the PR has to be marked as a breaking change (⚠️ ).

bnallapeta commented on controllers/openstackmachine_controller_test.go:
```diff
@@ -82,9 +82,6 @@ func TestOpenStackMachineSpecToOpenStackServerSpec(t *testing.T) {
 			SecurityGroups: []infrav1.SecurityGroupParam{
-				{
-					ID: ptr.To(openStackCluster.Status.WorkerSecurityGroup.ID),
-				},
```
This was due to the other change. Its fixed.

bnallapeta commented on docs/book/src/topics/hosted-control-plane.md:
````diff
@@ -0,0 +1,449 @@
+Obtain the tenant cluster kubeconfig:
+
+```bash
+kubectl -n hcp-system get secrets | grep -i kubeconfig
+kubectl -n hcp-system get secret <NAME> -o jsonpath='{.data.value}' | base64 -d > tenant.kubeconfig
````
Done.

bnallapeta commented on docs/book/src/topics/hosted-control-plane.md:
```diff
@@ -0,0 +1,449 @@
+
+This guide demonstrated how to deploy a Kubernetes cluster on OpenStack using a hosted control plane architecture with Cluster API Provider OpenStack (CAPO) and k0smotron. The key benefits of this approach include:
+
+### Advantages of Hosted Control Planes
+
+- **Resource Efficiency**: Control plane components run as pods on the management cluster, reducing the infrastructure footprint
+- **Simplified Management**: Centralized control plane management across multiple tenant clusters
+- **High Availability**: Leverages the management cluster's infrastructure for control plane resilience
+- **Cost Optimization**: Eliminates the need for dedicated control plane nodes in each tenant cluster
```
Done.

bnallapeta commented on docs/book/src/topics/hosted-control-plane.md:
````diff
@@ -0,0 +1,449 @@
+#### Hosted Control Plane Cluster Manifests (`openstack-hcp-cluster.yaml`)
+
+```yaml
+apiVersion: v1
+data:
+  cacert: null
+  clouds.yaml: <BASE64_OF_clouds.yaml_HERE>
+kind: Secret
+metadata:
+  name: openstack-cloud-config
+  namespace: hcp-system
+---
````
Done.

bnallapeta commented on docs/book/src/topics/hosted-control-plane.md:
````diff
@@ -0,0 +1,449 @@
+
+### 1. Prepare OpenStack Credentials
+
+Create a base64-encoded OpenStack configuration file. The cluster manifest expects a secret named `openstack-cloud-config` in the `hcp-system` namespace with a `clouds.yaml` key.
+
+Encode your `clouds.yaml` file:
+
+**Linux:**
+```bash
+base64 -w0 clouds.yaml > clouds.yaml.b64
+```
+
+**macOS:**
+```bash
+base64 -b0 clouds.yaml > clouds.yaml.b64
+```
+
+Copy the single-line contents of `clouds.yaml.b64` and paste it into the cluster manifest at `data.clouds.yaml` (replace `<BASE64_OF_clouds.yaml_HERE>`).
````
Done.

bnallapeta commented on docs/book/src/topics/hosted-control-plane.md:
````diff
@@ -0,0 +1,449 @@
+  ```
+- k0smotron installed (docs: https://docs.k0smotron.io/stable/install/#software-prerequisites)
+- A LoadBalancer implementation for the hosted control plane service (OpenStack CCM/Octavia or MetalLB)
+- OpenStack Cinder CSI driver installed and operational on the management cluster with:
+  - Valid OpenStack credentials secret in the `kube-system` namespace
+  - A default StorageClass configured (if you install the openstack-cinder-csi helm chart, then the StorageClass will be created for you)
````
Done.

bnallapeta commented:
@lentzi90 addressed all the comments on docs. Just one conversation yet to be resolved. 

Also, with this commit, I updated all references of tenant cluster to workload cluster to stay in touch with CAPI/CAPO nomenclature.

bnallapeta commented:
/remove-hold
